File: DataEngineering/01_Architecture_and_Storage/projects/golang/01_local_data_lake_simulation/main.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"path/filepath"
	"time"

	"github.com/xitongsys/parquet-go-source/local"
	"github.com/xitongsys/parquet-go/parquet"
	"github.com/xitongsys/parquet-go/writer"
)

// EventRecord defines the structure of our data lake events.
type EventRecord struct {
	ID        int64  `parquet:"name=id, type=INT64"`
	Event     string `parquet:"name=event, type=UTF8, encoding=PLAIN_DICTIONARY"`
	Timestamp int64  `parquet:"name=timestamp, type=INT64, logicaltype=TIMESTAMP_MILLIS"`
}

func main() {
	baseDir := "data_lake/raw"
	year, month, day := 2023, 10, 1
	partitionDir := filepath.Join(baseDir, fmt.Sprintf("year=%d/month=%02d/day=%02d", year, month, day))

	// Create partition directory structure
	// This would typically be handled by a storage layer (S3) or an abstraction like PyArrow
	// but here we manually create the directories for local simulation.
	
	// Create local Parquet file writer
	filePath := filepath.Join(partitionDir, "events.parquet")
	fw, err := local.NewLocalFileWriter(filePath)
	if err != nil {
		log.Fatalf("Failed to create local file: %v", err)
	}
	defer fw.Close()

	// Initialize Parquet writer
	pw, err := writer.NewParquetWriter(fw, new(EventRecord), 4)
	if err != nil {
		log.Fatalf("Failed to create Parquet writer: %v", err)
	}

	pw.RowGroupSize = 128 * 1024 * 1024 // 128 MB
	pw.CompressionType = parquet.CompressionCodec_SNAPPY

	events := []string{"login", "logout", "purchase", "view"}

	// Generate and write synthetic data
	for i := 0; i < 1000; i++ {
		record := EventRecord{
			ID:        int64(i),
			Event:     events[rand.Intn(len(events))],
			Timestamp: time.Now().UnixMilli(),
		}
		if err := pw.Write(record); err != nil {
			log.Fatalf("Write error: %v", err)
		}
	}

	if err := pw.WriteStop(); err != nil {
		log.Fatalf("WriteStop error: %v", err)
	}

	fmt.Println("Data lake simulation complete. Parquet data written to:", filePath)
}

File: DataEngineering/01_Architecture_and_Storage/projects/python/01_local_data_lake_simulation/main.py
import os
import random
from datetime import datetime, timedelta
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq

def generate_dummy_data(num_records: int = 100) -> pd.DataFrame:
    """Generates synthetic event data for the data lake."""
    data = []
    start_date = datetime(2023, 1, 1)
    for i in range(num_records):
        date = start_date + timedelta(days=random.randint(0, 365))
        record = {
            "id": i,
            "event": random.choice(["login", "logout", "purchase", "view"]),
            "timestamp": date.isoformat(),
            "year": date.year,
            "month": date.month,
            "day": date.day
        }
        data.append(record)
    return pd.DataFrame(data)

def ingest_to_data_lake(df: pd.DataFrame, base_dir: str = "data_lake/raw") -> None:
    """
    Ingests a Pandas DataFrame into a local Data Lake partitioned by year, month, and day.
    Utilizes PyArrow for efficient Parquet writing.
    """
    os.makedirs(base_dir, exist_ok=True)
    
    # Convert Pandas DataFrame to PyArrow Table
    table = pa.Table.from_pandas(df)
    
    # Write partitioned Parquet dataset
    pq.write_to_dataset(
        table,
        root_path=base_dir,
        partition_cols=['year', 'month', 'day'],
        compression='snappy',
        use_dictionary=True
    )

if __name__ == "__main__":
    print("Generating dummy dataset...")
    df = generate_dummy_data(1000)
    
    print("Ingesting data to local data lake (Parquet Format)...")
    ingest_to_data_lake(df)
    
    print("Data ingestion complete. Check the 'data_lake/raw' directory for partitioned Parquet files.")
